### PR TITLE
CC-7162: Add offset recovery and deterministic seeding

### PR DESCRIPTION
**Pull request by gharris1727**
This PR allows the Datagen connector to recover from source offsets in the event of failures, and generate data deterministically across those failures.

This depends on the release of an `avro-random-generator` version that contains a [determinism fix](https://github.com/confluentinc/avro-random-generator/pull/13`) and a [reentrant iterators feature](https://github.com/confluentinc/avro-random-generator/pull/14), which should either be `0.2.4` or `0.3.0`.

This feature adds one additional configuration option: `random.seed`, which allows for deterministic seeding of the random generator. All additional functionality is transparent to the user, and only affects the function of the connector under fault conditions.

The largest functional change is that after a datagen connector reaches the end of it's iteration and is then restarted, it will no longer start generating from the beginning of time. It will not generate any further records once it has reached the end.

### CODE REVIEW

**Comment by ybyzek** [comment]
@gharris1727 thanks for the PR.  cc: Connect team for review/approval.

Also, wrt

> It will not generate any further records once it has reached the end.

"Reached the end"...of what?  This occurs after how many records are sent and under what conditions?  One appeal of the datagen is that it can be configured with `-1` iterations which means it keeps going until stopped, so this appears to negate that behavior?

**Comment by gharris1727** [comment]
> "Reached the end"...of what? This occurs after how many records are sent and under what conditions? One appeal of the datagen is that it can be configured with `-1` iterations which means it keeps going until stopped, so does this PR negate that behavior?

@ybyzek Sorry for being vague. I meant the end of the iterations configuration. -1 is still respected as "infinity" and will continue generating records indefinitely.

The difference in behavior appears in the following situation:
1. The iterations configuration is set to 10
2. The connector generates 10 records and stops
3. The connector task is restarted

At this point, the old behavior would be to generate 10 more records.
The new behavior would not generate any further records.

**Comment by ybyzek** [comment]
@gharris1727 thanks for the clarification.  Sounds good.  

I'll defer on final review/approval to the connect engineering team, thank you!

File: src/main/java/io/confluent/kafka/connect/datagen/DatagenConnectorConfig.java [hunk 1]
```diff
@@ -38,6 +38,10 @@
   private static final String SCHEMA_KEYFIELD_DOC = "Name of field to use as the message key";
   public static final String QUICKSTART_CONF = "quickstart";
   private static final String QUICKSTART_DOC = "Name of quickstart to use";
+  public static final String RANDOM_SEED_CONF = "random.seed";
+  private static final String RANDOM_SEED_DOC = "Seed for generating random data. Two connectors "
+      + "started with the same seed will deterministically produce the same data. Each task will "
+      + "generate different data than the other tasks in the same connector.";
```
Comment (levzem) [inline src/main/java/io/confluent/kafka/connect/datagen/DatagenConnectorConfig.java]: might be good to mention the seed is just a number

File: pom.xml [hunk 2]
```diff
@@ -32,7 +32,7 @@
     <properties>
         <connect-runtime-version>2.0.0</connect-runtime-version>
         <confluent.version>5.1.0</confluent.version>
-        <confluent.avro.generator.version>0.2.0</confluent.avro.generator.version>
+        <confluent.avro.generator.version>0.2.4-SNAPSHOT</confluent.avro.generator.version>
```
Comment (levzem) [inline pom.xml]: is there a reason we can't wait for a released version or are you just issuing this PR preemptively?

Comment (gharris1727) [reply pom.xml]: I'm issuing this for review, and I don't intend to merge it until an ARG release has been made. Jenkins can't even build this :^)

**Comment by gharris1727** [comment]
@levzem 
> also maybe a test that it is indeed deterministic would be good?

The test that I include verifies the determinism of the connector after restarts, which should cover the functionality of the configuration option.

File: pom.xml (hunk 2 above)
Comment (C0urante) [reply pom.xml]: [gee, if only someone would publish a new release already...](https://repo1.maven.org/maven2/io/confluent/avro/avro-random-generator/0.3.0)

File: README.md [hunk 3]
```diff
@@ -120,7 +120,7 @@ Parameter | Description | Default
 -|-|-
 `kafka.topic` | Topic to write to | 
 `max.interval` | Max interval between messages (ms) | 500
-`iterations` | Number of messages to send, or less than 1 for unlimited | -1
+`iterations` | Number of messages to send from each task, or less than 1 for unlimited | -1
```
Comment (C0urante) [inline README.md]: Can we also update the `DatagenConnectorConfig.ITERATIONS_DOC` field with this?

Comment (gharris1727) [reply README.md]: this is the more important place to add them, oops

**Pull request merged**